### PR TITLE
Avoid ftell on null

### PR DESCRIPTION
**Pull request by tbeu**

### CODE REVIEW

**Review by clanmills — COMMENTED** [review]
Thomas:  Thanks for working on this.  I have my doubts about clang-tidy formatting.  When we "brutalise" the code with this, we introduce thousands of cosmetic changes.  We'll have difficulty applying security fixes to different editions of the code.

About 18 months ago, Luis proposed that we should adopt this and reformat all the code.  We agreed to only reformat around changes which were essential to fix issues or add new features.  Is it possible to revert to the original code, apply your change, and "clang tidy" the region of change?

Andreas wrote most of the Exiv2 code and I think it's well written and nicely laid out.  The v0.28 project is the "all new Exiv2".  As well as "modernising" the code to C++11, I think it's OK to reformat everything.  And use clang-tidy everywhere in future.  However for "the dots" (Exiv2 v0.27.1 etc...) can we avoid major cosmetic changes to the code?

Gilles (the owner of DigiKam) wanted use to use cppcheck.  Again, thousands of cosmetic changes to the code.  OK for v0.28?  Fine with me.  However on the dots, I prefer to leave things as they are.

I feel much the same about DataBuf.  It's been there for a long time.  However it's a real 1980's "C" kind of thing.  We can live with it for "the dots" and engineer something more robust/resilient/safer for v0.28.  I agree with Dan about discussing DataBuf on github.  We use Slack for one-to-one conversations and stuff that's way off topic (weather, vacations, family, politics).

I've been using Visual Studio since 1994 and really like it.  I use CLion on Mac and Linux and it works well.  I obtained 3 free/open-source CLion licenses for Luis, Dan and myself.  CLion has support for cmake, cppcheck and clang-tidy.  You can get a 30 day eval license for CLion and, if you like it, I'll request a license for you. 

I'm really pleased to see you re-surface and engage with the team.  Thanks for your participation.

File: src/basicio.cpp [hunk 1]
```diff
@@ -236,15 +236,14 @@ namespace Exiv2
             return 0;
         }
 
+        if (fp_ == nullptr)
```
Comment (D4N) [inline src/basicio.cpp]: This should be at the beginning of this function for consistency. Then the assert can go.

**Review by D4N — DISMISSED** [review]
Please reformat only `int FileIo::Impl::switchMode(OpMode opMode)` with clang format.

File: src/basicio.cpp (hunk 1 above)
Comment (tbeu) [reply src/basicio.cpp]: Agreed. Can do.

**Review by piponazo — DISMISSED** [review]

**Comment by D4N** [comment]
@tbeu This looks good, thanks!

@tbeu @piponazo Would you guys mind if we hijack this PR to try out what mergify can do? If yes, @tbeu could you push an arbitrary commit? Then mergify should in theory dismiss @piponazo's review.

**Comment by tbeu** [comment]
Yes and done.

**Comment by D4N** [comment]
> Alternatively, you can also check this one:

Where is that setting? If github supports that by themselves, then we don't need to rely on mergify for this.

Anyway, thanks @tbeu, at least mergify can dismiss reviews. Now we only need to teach it to properly backport commits. Feel free to drop the additional commit, I'll approve and merge this then.

**Comment by tbeu** [comment]
Click the edit button for the protected branches at https://github.com/Exiv2/exiv2/settings/branches.

**Review by piponazo — DISMISSED** [review]

**Comment by piponazo** [comment]
@D4N could yo retrigger the failing job on gitlab? Do you know how I could get access to re-run those jobs?

**Comment by D4N** [comment]
@piponazo I can give you access to the project on gitlab, but in this case it won't help, as the problem is the outdated `.gitlab-ci.yml`. @tbeu Needs to pull & rebase on top of master.

**Comment by tbeu** [comment]
> @tbeu Needs to pull & rebase on top of master.

Done.

**Review by D4N — APPROVED** [review]
Thanks for your patience and the experiments!

**Comment by D4N** [comment]
In case gitlab fails, merge this, it's a broken repository on CentOS.

**Pull request merged**